src/plugins/gtd_helper/index.tsx

```diff
@@ -0,0 +1,228 @@
+import * as React from 'react'; // tslint:disable-line no-unused-variable
+
+// tslint:disable-next-line:max-line-length
+import { registerPlugin, PluginApi } from '../../assets/ts/plugins';
+import { Logger } from '../../shared/utils/logger';
+import { matchWordRegex } from '../../assets/ts/utils/text';
+import Session from '../../assets/ts/session';
+import * as _ from 'lodash';
+import { Row, SerializedBlock } from '../../assets/ts/types';
+import Path from '../../assets/ts/path';
+
+// TODO:
+// - move all from _getChildren to getChildren which is based on path rather than row.
+// - use line renderHooks to automatically detect if GTD tag is present or not and clone it.
+// - Support link to original parent in gtd tasks.
+class GtdHelperPlugin {
+    private api: PluginApi;
+    private logger: Logger;
+    private gtdRegex: RegExp;
+    private loggerPrefix: string;
+    private gtdContainedRows: _.Dictionary<Set<any>>;
+    private clonedRows: Set<Row>;
+    private gtdNodeRow: Row;
+    private nodesUpdatedCounter: number;
+
+
+    constructor(api: PluginApi) {
+        this.api = api;
+        this.logger = this.api.logger;
+        this.gtdRegex = matchWordRegex('\\#\\w+?');
+        this.loggerPrefix = 'GTD Plugin: ';
+        this.gtdContainedRows = {};
+        this.clonedRows = new Set<Row>();
+        this.gtdNodeRow = -1;
+        this.nodesUpdatedCounter = 0;
+
+        const that = this;
+        this.api.registerAction(
+            'generate-gtd-tasks',
+            'Generates GTD tasks for the document',
+            async function ({ session }) {
+                await that.processGtdAction(session);
+            },
+        );
+
+
+        this.api.registerDefaultMappings(
+            'NORMAL',
+            {
+                'generate-gtd-tasks': [['ctrl+t']],
+            },
+        );
+    }
+
+    private log(message: any) {
+        this.logger.info(this.loggerPrefix + message.toString());
+    }
+
+    private async getChildWithText(row: Row, text: string): Promise<Row> {
+        if (await this.api.session.document.hasChildren(row)) {
+            let children = await this.api.session.document._getChildren(row);
+
+            let foundRow: Row = -1;
+            await Promise.all(children.map(async (child_row) => {
+                let child_text = await this.api.session.document.getText(child_row);
+                if (child_text.toLowerCase() === text.toLowerCase()) {
+                    foundRow = child_row;
+                }
+            }));
+
+            return foundRow;
+        }
+        return -1;
+    }
+
+    private async getGtdNode() {
+        this.gtdNodeRow = await this.getChildWithText(this.api.session.document.root.row, 'gtd');
```

Comment:
interesting - would it be more efficient/convenient to explicitly mark the gtd nodes and persist them?  there's storage meant for plugin use, `this.api.setData(key, value)`

Reply:
Yeah thought of storing gtdNodeRow and also all the clonedRows. But, if user ends up either deleting it or moving it somewhere might fail. quick question, does moving a node to another tree cause rowId to update?

Reply:
the operation "[dd] delete -> [p] paste" doesn't change the row ID.  "[y] yank" makes a new row ID though.  so both semantics are possible

Reply:
Hmm, then for the first version I'll keep this logic. Will figure out a better way to cache the rowIds rather than refetching all the time.

Was also thinking using listener for any line edit and automatically recompute GTD tree. "renderLine?" or something should be the one right? Or is there a better way to do it?

Reply:
i dont think there's an editLine listener.  2 options:
- in document.ts, just add something like `this.emit('setLine', { ... some_info })` in the function `public async setLine(row: Row, line: Line)`.  
- `renderLineTokenHook` might be an ok/slightly hacky replacement, since any edited line should re-render.  the hooks are like listeners, except that they can return a modified input, so your hook would just return the original input.

Reply:
another option is to explicitly mark the GTD tags, e.g. "gt" in normal mode would mean "get this done today!".  would be a bigger change though!